### PR TITLE
Parallel targets

### PR DESCRIPTION
**Pull request by mem48**
First crack at #258

image

Untangling everything will take a while. But the main change, splitting the commute routing into separate targets is done.

This now means that when using `tar_make_future` code can be running while waiting for routing from cyclestreets. 

There is a bit more splitting to do. E.g. the schools and some of the task at the end.  E.g. doing the zone stats and the school stats in the same target results in unnecessary waiting. It should also be possible to build the final rnets before the balanced routing is done as the balanced routes are not on the website.

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Good to see. Will take a look now.

**Comment by mem48** [comment]
Schools now also split 

image

**Comment by Robinlovelace** [comment]
There are pros and cons:

- Pro: a bit faster. Do we know by how much? Best we can hope for is from ~24 hours to ~8 hours assuming best case 4x speed up.
- Con: code complexity. Rather than setting the network types in `parameters$plans` these are now hardcoded. There are a lot more lines of code.
- Con: harder to debug and understand.

So I'm not convinced. We can get some of the benefits by reviving #78 .

**Comment by Robinlovelace** [comment]
Is there a way of doing this without adding so many extra lines of code?

**Comment by Robinlovelace** [comment]
I wonder if this could be a good use case of static branching: https://books.ropensci.org/targets/static.html#map

**Comment by mem48** [comment]
It won't be 3x faster. The build can never be faster than the routing and the routing is 75% of the build time.

But it solves a scheduling problem. There are two computers involved in the build. The build computer and the cyclestreets server. Currently only one is working at any moment. This parallelisation allows both to work simultaneously.

There are slightly more lines of code in `_targets.R`  750 vs 566 but I find it much easier to read. Most targets now run a single function, and many are a single line long. 

You could easily compress

```
tar_target(rnet_gq_school_balanced, {
  segments2rnet(rs_school_balanced[[1]]$segments)
}),
```
to a single line if the number of lines of code matters. There is also some further cleaning that can be done.

I think this is easier to debug, as smaller targets mean it's clearer where a failure has occurred.

It also means smaller datasets `tar_load(uptake_commute_fastest)` is 1/4 the size of `tar_load(uptake_list_commute)`

**Comment by mem48** [comment]
In a way, this is the only further speed improvement that is worth doing. As the routing is the slowest process any time that the cyclestreets server is idle slows down the critical path and so the whole build. Splitting up the work like this ensures that cyclestreets has very little downtime. 

Once this is implemented there is no need to make any other stage faster as they will all be done while waiting for cyclestreets to finish the routing.

The last thing that needs doing is splitting the combining rnets work so that it does not require the balanced routes.  This will mean that the pmtiles can be built before all the routing is completed which would mean maximum possible performance.

**Comment by Robinlovelace** [comment]
I see your points, although it does not move in the direction of automation. If we want to add/remove another plan type we have to manually comment out/remove/add lines of code, rather than just editing the config file.

Plan to approve after I see the results of this in any case.

Will test the latest version you have on a test with 250 and see what happens, and how much faster it is.

**Comment by Robinlovelace** [comment]
Testing it out now...

**Comment by Robinlovelace** [comment]
Current world view, is this really running on all cores @mem48 ?

image

**Comment by Robinlovelace** [comment]
Seems a bit quicker in sending the jobs: 

image

**Comment by mem48** [comment]
No it only uses two workers

```
future::plan(future::multisession, workers = 2)
```

**Comment by Robinlovelace** [comment]
How can we see if it's running in parallel from the output?

image

**Comment by mem48** [comment]
You can see that it is building rnets before all the routing is done

**Comment by mem48** [comment]
A that it is building rnets in 0.2 seconds. I think that is targets way of saying "retrieving the finished work from the worker only took 0.2 seconds"

**Comment by Robinlovelace** [comment]
Maybe I need to `tar_invalidate()`. Will try again...

**Comment by Robinlovelace** [comment]
Just running 

```r
targets::tar_watch()
```

**Comment by Robinlovelace** [comment]
Going to re-run starting with this:

image

**Comment by Robinlovelace** [comment]
Currently stuck on 1st routing job it seems... 
image

**Comment by Robinlovelace** [comment]
Will document how long it takes + and speed-up...

**Comment by Robinlovelace** [comment]
Parallel version took just under 10 minutes

```
▶ end pipeline [9.815 minutes]
```

**Comment by Robinlovelace** [comment]
Trying on the main branch..

**Comment by Robinlovelace** [comment]
image

**Comment by Robinlovelace** [comment]
Just under 11 minutes. So ~10% faster with 250 routes. Shall we try with a larger build? I think so.

```
▶ end pipeline [10.903 minutes]
```

**Comment by mem48** [comment]
Yes, the larger the build the bigger the difference but I wouldn't expect more than 20%.

**Comment by Robinlovelace** [comment]
:+1: to separating them out but I don't see the link between this PR that parallelises the build and removing balanced routes which is a separate issue. On that note I've created #264 and an associated PR: #265 the changes to allow the build to pass should be smaller on that branch than this one.

**Comment by mem48** [comment]
I thought we agreed with sustrans that balanced routes should be part of the data just not in the website?

So we should still include the balanced routes and rnet, but not make them a prerequisite of making the combined rnet and pmtiles.

**Comment by Robinlovelace** [comment]
Fine by me.

**Comment by Robinlovelace** [comment]
Currently trying this on Windows:

image

**Comment by Robinlovelace** [comment]
Update and FYI @mem48 it's now done 2 sets of school routes and is at 50% memory use. I wonder if there's a way of removing all memory after a target has finished:


image

**Comment by Robinlovelace** [comment]
OK, I see garbage collection is on:

https://github.com/nptscot/npt/blob/46b47fb9238b279ce3c7ffd89453ae9837179e2d/_targets.R#L20

**Comment by Robinlovelace** [comment]
Which would help explain why memory use has now dropped down again, to 18 GB.

**Comment by Robinlovelace** [comment]
Heads-up @mem48 it failed with the following:

```
▶ start target rs_commute_fastest
✖ error target rs_commute_fastest
▶ end pipeline [1.665 hours]
Error:
! Error running targets::tar_make_future()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: Timeout was reached: [api.cyclestreets.net] Operation timed out after 60008 milliseconds with 0 bytes received
```

**Comment by Robinlovelace** [comment]
I think it's an R setting that auto time-outs after 1hr:

```r
options("timeout") 
# 60
```

**Comment by Robinlovelace** [comment]
Going to update timeout. Current status.

image

**Comment by Robinlovelace** [comment]
Relevant issue: https://github.com/ropensci/rtweet/issues/229

**Comment by mem48** [comment]
I've pushed updated that splits out a lot of the late stage work related to zone stats

**Comment by mem48** [comment]
Your error does not seem to be related to the PR, but a problem with your connection to cyclestreets.

**Comment by mem48** [comment]
Could be connected to #261

**Comment by Robinlovelace** [comment]
Progressing again now, I think... will check

**Comment by Robinlovelace** [comment]
@mem48 it failed with the following:

```
✖ error target rs_commute_fastest
▶ end pipeline [3.451 hours]
Error:
! Error running targets::tar_make_future()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: no applicable method for 'group_by' applied to an object of class "NULL"
```

**Comment by mem48** [comment]
Have you cleared the old data files? `group_by` is in `get_routes` but I can't see why it would fail on a large batch and not a small batch.

**Comment by Robinlovelace** [comment]
See #272

**Comment by mem48** [comment]
Updated `_targets.R` should reduce peak memory use by not loading two sets of routes into memory at the same time.

I made a chart to explain the workflow as the default layout is unclear.

image

File: _targets.R [hunk 1]
```diff
@@ -223,8 +225,12 @@ tar_target(rs_school_fastest, {
   rs
 }),
 
+tar_target(done_school_fastest, {
+  exists(rs_school_fastest) #Hack for scheduling
```
Comment (Robinlovelace) [inline _targets.R:229]: This will still load into memory the whole object. Worth opening an issue in the targets issue tracker IMO.

**Comment by Robinlovelace** [comment]
> Updated `_targets.R` should reduce peak memory use by not loading two sets of routes into memory at the same time.
> 
> I made a chart to explain the workflow as the default layout is unclear.
> 
> image

Looks good.

File: _targets.R [hunk 2]
```diff
@@ -226,7 +226,7 @@ tar_target(rs_school_fastest, {
 }),
 
 tar_target(done_school_fastest, {
-  exists(rs_school_fastest) #Hack for scheduling
+  exists("rs_school_fastest") #Hack for scheduling
```
[the remark this answers is not in the document]
Comment (mem48) [reply _targets.R]: `exists()` only takes a character input i.e the name of the object, not the object itself. But if you quote the object name then targets does not recognise it for creating the dependency graph.  https://stackoverflow.com/questions/69662145/r-targets-package-how-to-refer-to-an-existing-target-using-a-string So I've gone back to using length.

Should still produce the desired result of lower peak memory use, but perhaps not a fast as I hoped.

**Comment by mem48** [comment]
National Build Running

image

using `tar_watch` before `tar_make_future` shows a cool real-time update. You can see that all the school's post-processing is done before the first batch of commute routing has been completed.

**Comment by mem48** [comment]
Currently with 4 workers and using a very manageable 3GB of RAM, but not working very hard at the moment because waiting for routes

**Comment by Robinlovelace** [comment]
Good to hear and fingers crossed. FYI @mem48 I finally checked on my laptop and it did fail, with this message:

```
▶ end pipeline [4.591 hours]
Error:
! Error running targets::tar_make_future()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: cannot allocate vector of size 158.1 Mb
```

**Comment by mem48** [comment]
Peak memory post commuter routing is 81 GB
image

Not that 79GB is on the single worker doing the routing. So this is not a parallel problem.

**Comment by Robinlovelace** [comment]
The switch to pulling down both segment and route level geometries increased memory use.

**Comment by mem48** [comment]
Full build complete, high memory usage is due to large batch sizes now used in cyclestreets

**Comment by Robinlovelace** [comment]
:tada:

File: R/zone_stats.R [hunk 3]
```diff
@@ -1,150 +1,131 @@
...
-#' @param schl uptake_list_school list
-#' @param zones sf data frame
+#' @param nm name to append to columns
 #' 
 #' the names are appended to column names 
-uptake_to_zone_stats <- function(comm, schl, zones){
-  
-  #comm = uptake_list_commute
-  #schl = uptake_list_school
+make_commute_stats <- function(comm, nm){
```
Comment (Robinlovelace) [inline R/zone_stats.R:6]: Should be `=` to be consistent with the project's style.

File: R/zone_stats.R [hunk 4]
```diff
@@ -1,150 +1,131 @@
...
                                 orig_bicycle_ebike = sum(bicycle_ebike, na.rm = TRUE),
                                 orig_car_ebike = sum(car_ebike, na.rm = TRUE),
                                 orig_public_transport_ebike = sum(public_transport_ebike, na.rm = TRUE),
                                 orig_foot_ebike = sum(foot_ebike, na.rm = TRUE),
+                                orig_taxi_ebike = sum(taxi_ebike, na.rm = TRUE),
                                 
                                 orig_quietness = mean(quietness),
                                 orig_route_hilliness = mean(route_hilliness, na.rm = TRUE)
   )
-  comm_from <- tidyr::pivot_wider(comm_from,
```
Comment (Robinlovelace) [inline R/zone_stats.R:62]: Glad to see all this hardcoded and verbose code removed.

File: R/zone_stats.R [hunk 5]
```diff
@@ -1,150 +1,131 @@
...
+                              foot_ebike = sum(foot_ebike, na.rm = TRUE),
+                              other_ebike = sum(other_ebike, na.rm = TRUE),
+                              
+                              quietness = mean(quietness, na.rm = TRUE),
+                              route_hilliness = mean(route_hilliness, na.rm = TRUE)
+  )
+  
+  schl_to <- dplyr::ungroup(schl_to)
+  
+  schl_to <- tidyr::pivot_wider(schl_to,
```
Comment (Robinlovelace) [inline R/zone_stats.R:99]: All this is a bit clunky. Any way to neaten it up?

File: R/zone_stats.R [hunk 6]
```diff
@@ -393,4 +234,93 @@ export_zone_json <- function(x,  idcol = "DataZone", path = "outputdata/json", z
 
 
 
-
+#' Function to summarise routes into zone statistics
```
Comment (Robinlovelace) [inline R/zone_stats.R:237]: Maybe move this to `code/tests` if not useful?

File: _targets.R [hunk 7]
```diff
@@ -1,40 +1,43 @@
...
-# Follow the comments below to fill in this target script.
-# Then follow the manual to check and run the pipeline:
-#   https://books.ropensci.org/targets/walkthrough.html#inspect-the-pipeline # nolint
+# Instructions
+# 1) library(targets)
+# 2) Optional - to see real-time updates of progress
+# tar_watch(seconds = 60, targets_only = TRUE)
+# 3) To run the build
+# tar_make_future(workers = 4)
+# If your RAM limited use tar_make() to run one job at a time
```
Comment (Robinlovelace) [inline _targets.R:7]: :+1:

File: _targets.R [hunk 8]
```diff
@@ -1,40 +1,43 @@
...
+  remotes::install_github("dabreegster/odjitter", subdir = "r")
+  remotes::install_github("ropensci/stplanr")# Improved overline
+  remotes::install_cran("targets")
+}
+
 library(targets)
-library(tidyverse)
-# library(tmap)
-library(stplanr)
+library(magrittr) # Light load of %>%
```
Comment (Robinlovelace) [inline _targets.R:18]: We currently use 2 pipe operators. We should standardise on `|>`.

**Review by Robinlovelace — COMMENTED** [review]
Looking good. See minor comments above. I'm going to reproduce the results with a small build to approving. Overall, great work Malcolm!

**Comment by Robinlovelace** [comment]
Going to try with 250 routes...

**Comment by Robinlovelace** [comment]
Also @mem48 can you share the pmtiles from the results? Will be good to check locally, I don't see them in the release.

**Comment by Robinlovelace** [comment]
Another issue with the release @mem48: the secondary network seems way smaller than the primary network. Is this expected?

image

**Comment by Robinlovelace** [comment]
A weird thing with the build pipeline, it seems that the save outputs target can complete before rnet and even some routing targets are complete:

image

Could that explain some of the small outputs?

**Comment by Robinlovelace** [comment]
This can be seen in the dependency graph where the commute stats target for example happens 'after' the save targets target:

image

**Comment by Robinlovelace** [comment]
Other pathway of same image showing that save outputs does not depend on all the rnets being complete.

image

**Comment by mem48** [comment]
Need to replace `length(done_commute_balanced)` with `length(rnet_commute_balanced)`

**Review by Robinlovelace — CHANGES_REQUESTED** [review]
Some things not quite right with this, outputs in release have variable sizes for rnet files and save outputs target should always go after rnet targets.

**Comment by Robinlovelace** [comment]
> Need to replace `length(done_commute_balanced)` with `length(rnet_commute_balanced)` in `upload_data`

OK, you happy to give it a bash? Good progress in any case.

**Comment by mem48** [comment]
Done

**Comment by Robinlovelace** [comment]
OK, I'll try reproducing the results again.

**Comment by Robinlovelace** [comment]
Did that affect the full build? Worth re-running your side also I think so we have complete outputs and pmtiles to check.

**Comment by Robinlovelace** [comment]
Have you double checked 

```r
targets::tar_visnetwork(T)
```

After the change above @mem48 ?

**Comment by Robinlovelace** [comment]
This is what I see after pulling down the latest changes. Not ready yet I think... Please double check and re-request review when good to go and I'll take a proper look and reproduce the results.

image

**Comment by mem48** [comment]
> This is what I see after pulling down the latest changes. Not ready yet I think... Please double check and re-request review when good to go and I'll take a proper look and reproduce the results.

THis looks fine to me, wht is the problem?

**Comment by Robinlovelace** [comment]
> THis looks fine to me, wht is the problem?

`rnet_commute_balanced` generated after save outputs.

**Comment by mem48** [comment]
Save outputs does not use `rnet_commute_balanced`

**Comment by Robinlovelace** [comment]
Aha, and the target could probably be commented out.

**Comment by Robinlovelace** [comment]
Can you share the pmtiles resulting from the full build?

**Comment by mem48** [comment]
On the website already

**Comment by mem48** [comment]
I spotted one bug that the values are not rounded, now fixed in code

**Comment by Robinlovelace** [comment]
> I spotted one bug that the values are not rounded, now fixed in code

Glad the review is helpful. Can you generate the pmtiles for comparison with previous build? Will check when ready, off to park.

**Comment by mem48** [comment]
Updated pmtiles now on server

**Comment by mem48** [comment]
image

**Comment by Robinlovelace** [comment]
Fantastic, will take a look at the test website now.

**Comment by Robinlovelace** [comment]
Have updated the date here: https://github.com/nptscot/nptscot.github.io/commit/477cb6cb7c78bdad78f399d2a69fe3c9bc4dfa73

**Review by Robinlovelace — APPROVED** [review]

**Comment by Robinlovelace** [comment]
Merged! Also trying a build here for reproducibility's sake, let's see what happens..

**Comment by Robinlovelace** [comment]
Chugging away fine with 1 worker on laptop with 64 GB RAM FYI @mem48, the various tweaks, particularly the `routes_done` hack, seem to have worked.

**Pull request merged**